File: scripts/validate_skill_anatomy.py
#!/usr/bin/env python3
"""Validate Agent Booster Pack skill files and plugin mirror drift."""

from __future__ import annotations

import argparse
import json
import re
import shutil
import sys
import tempfile
from dataclasses import dataclass
from json import JSONDecodeError
from pathlib import Path
from typing import Any


REQUIRED_SECTIONS = (
    "When to Use",
    "When NOT to Use",
    "Verification",
)
MAX_DESCRIPTION_LENGTH = 200

NAME_RE = re.compile(r"^name:\s+[a-z][a-z0-9-]*\s*$", re.MULTILINE)
DESCRIPTION_RE = re.compile(r"^description:", re.MULTILINE)
ATTRIBUTION_RE = re.compile(r"\b[Pp]er\s+(?:[A-Z][a-z]+\s+)?[A-Z][a-z]+\b")
TRIPWIRE_HEADER_RE = re.compile(
    r"^\|\s*Trigger\s*\|\s*Do this instead\s*\|\s*False alarm\s*\|\s*$",
    re.MULTILINE,
)


@dataclass(frozen=True)
class SkillFinding:
    """Validation findings for a single SKILL.md file."""

    path: Path
    problems: tuple[str, ...]


def default_skills_dir() -> Path:
    """Return the default canonical skills directory."""
    return Path(__file__).resolve().parent.parent / "agents" / ".agents" / "skills"


def section_re(name: str) -> re.Pattern[str]:
    """Build a section-heading matcher for a required heading."""
    return re.compile(rf"^##+\s+{re.escape(name)}\s*$", re.MULTILINE)


def body_without_reference_sections(body: str) -> str:
    """Remove lines under References or Canon before attribution checks."""
    kept: list[str] = []
    in_reference_section = False

    for line in body.splitlines():
        if re.match(r"^##+ (References|Canon)\s*$", line):
            in_reference_section = True
            continue
        if not in_reference_section:
            kept.append(line)

    return "\n".join(kept)


def frontmatter_description(head: str) -> str | None:
    """Return the normalized frontmatter description when present."""
    lines = head.splitlines()
    description: list[str] = []
    in_description = False

    for line in lines:
        if line.startswith("description:"):
            in_description = True
            value = line.split(":", 1)[1].strip()
            if value and value not in {">", ">-", "|", "|-"}:
                description.append(value)
            continue

        if in_description:
            if line.startswith("  "):
                description.append(line.strip())
                continue
            break

    if not in_description:
        return None

    return " ".join(description)


def validate_skill_file(path: Path) -> SkillFinding | None:
    """Validate one SKILL.md file and return findings when it fails."""
    body = path.read_text(encoding="utf-8")
    head = "\n".join(body.splitlines()[:30])
    problems: list[str] = []

    if not NAME_RE.search(head):
        problems.append("frontmatter missing name or not kebab-case")
    description = frontmatter_description(head)
    if description is None:
        problems.append("frontmatter missing description")
    elif len(description) > MAX_DESCRIPTION_LENGTH:
        problems.append(
            "frontmatter description too long "
            f"({len(description)} > {MAX_DESCRIPTION_LENGTH} characters)"
        )

    for heading in REQUIRED_SECTIONS:
        if not section_re(heading).search(body):
            problems.append(f"missing section: ## {heading}")

    if section_re("Common Rationalizations").search(body):
        problems.append(
            "obsolete section: ## Common Rationalizations -- use ## Tripwires"
        )
    if section_re("Red Flags").search(body):
        problems.append("obsolete section: ## Red Flags -- fold into ## Tripwires")
    if re.search(r"^\|\s*Excuse\s*\|\s*Reality\s*\|", body, re.MULTILINE):
        problems.append(
            "obsolete table header: use Trigger | Do this instead | False alarm"
        )
    if section_re("Tripwires").search(body) and not TRIPWIRE_HEADER_RE.search(body):
        problems.append(
            "Tripwires table must use: Trigger | Do this instead | False alarm"
        )

    attribution_lines = [
        line
        for line in body_without_reference_sections(body).splitlines()
        if not line.startswith(">") and ATTRIBUTION_RE.search(line)
    ]
    if attribution_lines:
        problems.append("inline 'per <expert>' attribution found -- move to References")

    if not problems:
        return None

    return SkillFinding(path=path, problems=tuple(problems))


def validate_skills(skills_dir: Path) -> list[SkillFinding]:
    """Validate all SKILL.md files under a skills directory."""
    findings: list[SkillFinding] = []
    for skill_file in sorted(skills_dir.glob("*/SKILL.md")):
        finding = validate_skill_file(skill_file)
        if finding is not None:
            findings.append(finding)
    return findings


def print_skill_findings(skills_dir: Path, findings: list[SkillFinding]) -> None:
    """Print skill anatomy findings in the historical script format."""
    for finding in findings:
        rel = finding.path.relative_to(skills_dir)
        print(rel)
        for problem in finding.problems:
            print(f"  - {problem}")

    if findings:
        print()
        print(f"{len(findings)} skill(s) failed anatomy validation")
    else:
        print("all skills conform to the anatomy")


def plugin_skills_dir(skills_dir: Path) -> Path:
    """Return the plugin skills directory for a canonical skills directory."""
    return skills_dir.parent.parent.parent / "plugin" / "skills"


def repo_root_for_skills_dir(skills_dir: Path) -> Path:
    """Return the repository root implied by a canonical skills directory."""
    return skills_dir.parent.parent.parent


def validate_plugin_drift(skills_dir: Path) -> int:
    """Check that plugin/skills mirrors canonical skill directories."""
    plugin_dir = plugin_skills_dir(skills_dir)
    if not plugin_dir.is_dir():
        return 0

    drift = 0
    for skill_dir in sorted(path for path in skills_dir.iterdir() if path.is_dir()):
        mirror = plugin_dir / skill_dir.name
        if not mirror.is_dir():
            print(
                f"plugin drift: {mirror} missing -- "
                "run scripts/generate_plugin_symlinks.py"
            )
            drift += 1
            continue

        for source_file in sorted(
            path for path in skill_dir.rglob("*") if path.is_file()
        ):
            relative = source_file.relative_to(skill_dir)
            mirror_file = mirror / relative
            if not mirror_file.is_file():
                print(f"plugin drift: {mirror_file} missing")
                drift += 1
                continue
            if source_file.read_bytes() != mirror_file.read_bytes():
                print(f"plugin drift: {mirror_file} differs from {source_file}")
                drift += 1

        for mirror_file in sorted(path for path in mirror.rglob("*") if path.is_file()):
            relative = mirror_file.relative_to(mirror)
            source_file = skill_dir / relative
            if not source_file.is_file():
                print(f"plugin drift: {mirror_file} has no canonical source")
                drift += 1

    for mirror in sorted(path for path in plugin_dir.iterdir() if path.is_dir()):
        if not (skills_dir / mirror.name).is_dir():
            print(f"plugin drift: {mirror} has no canonical skill")
            drift += 1

    if drift == 0:
        print("plugin/ skill mirror in sync with source")
    else:
        print()
        print(f"{drift} plugin mirror difference(s) found")

    return drift


def read_json_object(path: Path) -> tuple[dict[str, Any] | None, str | None]:
    """Read a JSON object or return a validation problem."""
    if not path.is_file():
        return None, f"missing {path}"

    try:
        data = json.loads(path.read_text(encoding="utf-8"))
    except JSONDecodeError as error:
        return None, f"{path} is not valid JSON: {error.msg}"

    if not isinstance(data, dict):
        return None, f"{path} must contain a JSON object"

    return data, None


def first_plugin_entry(marketplace: dict[str, Any]) -> dict[str, Any] | None:
    """Return the ABP plugin entry from a marketplace document."""
    plugins = marketplace.get("plugins")
    if not isinstance(plugins, list):
        return None

    for entry in plugins:
        if isinstance(entry, dict) and entry.get("name") == "abp":
            return entry

    return None


def validate_codex_plugin_package(skills_dir: Path) -> int:
    """Validate Codex plugin manifest and repo marketplace metadata."""
    root = repo_root_for_skills_dir(skills_dir)

    problems: list[str] = []
    marketplace_path = root / ".agents" / "plugins" / "marketplace.json"
    manifest_path = root / "plugin" / ".codex-plugin" / "plugin.json"

    marketplace, problem = read_json_object(marketplace_path)
    if problem is not None:
        problems.append(problem)
    elif marketplace is not None:
        if marketplace.get("name") != "abp":
            problems.append(f"{marketplace_path} name must be 'abp'")
        interface = marketplace.get("interface")
        if not isinstance(interface, dict):
            problems.append(f"{marketplace_path} interface must be an object")
        elif interface.get("displayName") != "Agent Booster Pack":
            problems.append(
                f"{marketplace_path} interface.displayName must be 'Agent Booster Pack'"
            )

        entry = first_plugin_entry(marketplace)
        if entry is None:
            problems.append(f"{marketplace_path} must include an 'abp' plugin entry")
        else:
            source = entry.get("source")
            if not isinstance(source, dict):
                problems.append(f"{marketplace_path} abp.source must be an object")
            else:
                if source.get("source") != "local":
                    problems.append(
                        f"{marketplace_path} abp.source.source must be 'local'"
                    )
                if source.get("path") != "./plugin":
                    problems.append(
                        f"{marketplace_path} abp.source.path must be './plugin'"
                    )

            policy = entry.get("policy")
            if not isinstance(policy, dict):
                problems.append(f"{marketplace_path} abp.policy must be an object")
            else:
                if policy.get("installation") != "AVAILABLE":
                    problems.append(
                        f"{marketplace_path} abp.policy.installation must be "
                        "'AVAILABLE'"
                    )
                if policy.get("authentication") != "ON_INSTALL":
                    problems.append(
                        f"{marketplace_path} abp.policy.authentication must be "
                        "'ON_INSTALL'"
                    )

            if entry.get("category") != "Coding":
                problems.append(f"{marketplace_path} abp.category must be 'Coding'")

    manifest, problem = read_json_object(manifest_path)
    if problem is not None:
        problems.append(problem)
    elif manifest is not None:
        if manifest.get("name") != "abp":
            problems.append(f"{manifest_path} name must be 'abp'")
        if manifest.get("skills") != "./skills/":
            problems.append(f"{manifest_path} skills must be './skills/'")

        interface = manifest.get("interface")
        if not isinstance(interface, dict):
            problems.append(f"{manifest_path} interface must be an object")
        else:
            if interface.get("displayName") != "Agent Booster Pack":
                problems.append(
                    f"{manifest_path} interface.displayName must be "
                    "'Agent Booster Pack'"
                )
            if interface.get("category") != "Coding":
                problems.append(f"{manifest_path} interface.category must be 'Coding'")

            capabilities = interface.get("capabilities")
            if not isinstance(capabilities, list) or not {
                "Read",
                "Write",
            }.issubset(set(capabilities)):
                problems.append(
                    f"{manifest_path} interface.capabilities must include "
                    "'Read' and 'Write'"
                )

            default_prompt = interface.get("defaultPrompt")
            if not isinstance(default_prompt, list) or len(default_prompt) > 3:
                problems.append(
                    f"{manifest_path} interface.defaultPrompt must contain "
                    "at most 3 prompts"
                )

        claude_marketplace, claude_problem = read_json_object(
            root / ".claude-plugin" / "marketplace.json"
        )
        if claude_problem is None and claude_marketplace is not None:
            claude_version = claude_marketplace.get("metadata", {}).get("version")
            if manifest.get("version") != claude_version:
                problems.append(
                    f"{manifest_path} version must match "
                    ".claude-plugin/marketplace.json metadata.version"
                )
            claude_entry = first_plugin_entry(claude_marketplace)
            if claude_entry is None:
                problems.append(
                    ".claude-plugin/marketplace.json must include an 'abp' plugin entry"
                )
            elif claude_entry.get("source") != "./plugin":
                problems.append(
                    ".claude-plugin/marketplace.json abp.source must be './plugin'"
                )

    if problems:
        for problem in problems:
            print(f"codex plugin: {problem}")
        print()
        print(f"{len(problems)} codex plugin package problem(s)")
    else:
        print("codex plugin package valid")

    return len(problems)


def write_fixture(path: Path, text: str) -> None:
    """Create a fixture SKILL.md file."""
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(text, encoding="utf-8")


def run_self_test() -> int:
    """Run validator self-tests against temporary good and bad skill fixtures."""
    tmp = Path(tempfile.mkdtemp())
    try:
        write_fixture(
            tmp / "good" / "SKILL.md",
            """---
name: good
description: Ok
---

# Good

## When to Use
- trigger
- one change per commit

## When NOT to Use
- other

## Verification
- [ ] check

## Tripwires
| Trigger | Do this instead | False alarm |
|---|---|---|
| "Should work" | Run the check. | Research notes that do not claim completion. |
""",
        )
        write_fixture(
            tmp / "bad" / "SKILL.md",
            """---
name: bad
description: >-
  This description is intentionally much too long for the skill routing surface
  because Codex and other agents load all skill descriptions before deciding
  which full skill body to request, so verbose trigger prose creates avoidable
  context pressure.
---

# Bad

Per Rich Hickey, prefer values over places.

## Overview

Stuff.

## Common Rationalizations
| Excuse | Reality |
|---|---|
| "Should work" | It might not. |

## Red Flags
- "Probably fine"
""",
        )

        findings = validate_skills(tmp)
        rendered = "\n".join(
            [str(finding.path.relative_to(tmp)) for finding in findings]
            + [problem for finding in findings for problem in finding.problems]
        )
        for expected in (
            "bad/SKILL.md",
            "When to Use",
            "Verification",
            "description too long",
            "per <expert>",
            "Common Rationalizations",
            "Red Flags",
            "Trigger | Do this instead | False alarm",
        ):
            if expected not in rendered:
                print(
                    f"self-test failed: missing {expected!r} in output",
                    file=sys.stderr,
                )
                print(rendered, file=sys.stderr)
                return 1

        good_path = Path("good/SKILL.md")
        if any(finding.path.relative_to(tmp) == good_path for finding in findings):
            print("self-test failed: good skill flagged", file=sys.stderr)
            print(rendered, file=sys.stderr)
            return 1

        print("self-test ok")
        return 0
    finally:
        shutil.rmtree(tmp)


def parse_args(argv: list[str]) -> argparse.Namespace:
    """Parse command-line arguments."""
    parser = argparse.ArgumentParser(
        description="Validate SKILL.md frontmatter, required sections, and plugin drift."
    )
    parser.add_argument(
        "skills_dir",
        nargs="?",
        type=Path,
        default=default_skills_dir(),
        help="skills directory; defaults to agents/.agents/skills",
    )
    parser.add_argument(
        "--self-test",
        action="store_true",
        help="run validator self-tests instead of validating the repository",
    )
    return parser.parse_args(argv)


def main(argv: list[str] | None = None) -> int:
    """Run the skill anatomy validator."""
    args = parse_args(sys.argv[1:] if argv is None else argv)

    if args.self_test:
        return run_self_test()

    skills_dir = args.skills_dir
    if not skills_dir.is_dir():
        print(f"not a directory: {skills_dir}", file=sys.stderr)
        return 2

    findings = validate_skills(skills_dir)
    print_skill_findings(skills_dir, findings)
    drift = validate_plugin_drift(skills_dir)
    codex_plugin_problems = validate_codex_plugin_package(skills_dir)

    return 1 if findings or drift or codex_plugin_problems else 0


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/validate-skill-anatomy.mjs
#!/usr/bin/env node
// Validate Agent Booster Pack skill files and plugin mirror drift.

import {
  cpSync,
  existsSync,
  mkdirSync,
  mkdtempSync,
  readdirSync,
  readFileSync,
  rmSync,
  statSync,
  writeFileSync,
} from "node:fs";
import { dirname, join, relative, resolve } from "node:path";
import { tmpdir } from "node:os";
import { fileURLToPath } from "node:url";

const REQUIRED_SECTIONS = ["When to Use", "When NOT to Use", "Verification"];
const MAX_DESCRIPTION_LENGTH = 120;
const MAX_TOTAL_DESCRIPTION_LENGTH = 2000;

const NAME_RE = /^name:\s+[a-z][a-z0-9-]*\s*$/m;
const ATTRIBUTION_RE = /\b[Pp]er\s+(?:[A-Z][a-z]+\s+)?[A-Z][a-z]+\b/;
function scriptDir() {
  return dirname(fileURLToPath(import.meta.url));
}

function defaultSkillsDir() {
  return resolve(scriptDir(), "../agents/.agents/skills");
}

function sectionRe(name) {
  return new RegExp(`^##+\\s+${name.replace(/[.*+?^${}()|[\]\\]/g, "\\$&")}\\s*$`, "m");
}

function skillSections(body) {
  const lines = body.split(/\r?\n/);
  const headings = lines
    .map((line, index) => ({ line, index }))
    .filter(({ line }) => /^##\s+/.test(line));

  return headings.map(({ line, index }, headingIndex) => {
    const next = headings[headingIndex + 1]?.index ?? lines.length;
    return { name: line.replace(/^##\s+/, "").trim(), lineCount: next - index - 1 };
  });
}

function tripwiresTooLong(body) {
  const sections = skillSections(body);
  const tripwires = sections.find((section) => section.name === "Tripwires");
  if (!tripwires) return false;

  const otherMax = Math.max(...sections.filter((section) => section.name !== "Tripwires").map((section) => section.lineCount));
  return tripwires.lineCount >= otherMax;
}

function bodyWithoutReferenceSections(body) {
  const kept = [];
  let inReferenceSection = false;
  for (const line of body.split(/\r?\n/)) {
    if (/^##+ (References|Canon)\s*$/.test(line)) {
      inReferenceSection = true;
      continue;
    }
    if (!inReferenceSection) kept.push(line);
  }
  return kept.join("\n");
}

export function frontmatterDescription(head) {
  const lines = head.split(/\r?\n/);
  const description = [];
  let inDescription = false;

  for (const line of lines) {
    if (line.startsWith("description:")) {
      inDescription = true;
      const value = line.split(":", 2)[1].trim();
      if (value && !new Set([">", ">-", "|", "|-"]).has(value)) {
        description.push(value);
      }
      continue;
    }

    if (inDescription) {
      if (line.startsWith("  ")) {
        description.push(line.trim());
        continue;
      }
      break;
    }
  }

  return inDescription ? description.join(" ") : null;
}

export function validateSkillFile(path) {
  const body = readFileSync(path, "utf8");
  const head = body.split(/\r?\n/).slice(0, 30).join("\n");
  const problems = [];

  if (!NAME_RE.test(head)) problems.push("frontmatter missing name or not kebab-case");
  const description = frontmatterDescription(head);
  if (description == null) {
    problems.push("frontmatter missing description");
  } else if (description.length > MAX_DESCRIPTION_LENGTH) {
    problems.push(`frontmatter description too long (${description.length} > ${MAX_DESCRIPTION_LENGTH} characters)`);
  }

  for (const heading of REQUIRED_SECTIONS) {
    if (!sectionRe(heading).test(body)) problems.push(`missing section: ## ${heading}`);
  }

  if (sectionRe("Common Rationalizations").test(body)) {
    problems.push("obsolete section: ## Common Rationalizations -- use ## Tripwires");
  }
  if (sectionRe("Red Flags").test(body)) {
    problems.push("obsolete section: ## Red Flags -- fold into ## Tripwires");
  }
  if (/^\|\s*Excuse\s*\|\s*Reality\s*\|/m.test(body)) {
    problems.push("obsolete table header: use positive Tripwires bullets");
  }
  if (tripwiresTooLong(body)) {
    problems.push("Tripwires must not be the longest section -- keep only high-probability skip/avoidance failures");
  }

  const attributionLines = bodyWithoutReferenceSections(body)
    .split(/\r?\n/)
    .filter((line) => !line.startsWith(">") && ATTRIBUTION_RE.test(line));
  if (attributionLines.length > 0) {
    problems.push("inline 'per <expert>' attribution found -- move to References");
  }

  return problems.length > 0 ? { path, problems } : null;
}

export function validateSkills(skillsDir) {
  const findings = [];
  let totalDescriptionLength = 0;

  for (const name of readdirSync(skillsDir).sort()) {
    const skillFile = join(skillsDir, name, "SKILL.md");
    if (!existsSync(skillFile)) continue;
    const body = readFileSync(skillFile, "utf8");
    const head = body.split(/\r?\n/).slice(0, 30).join("\n");
    const description = frontmatterDescription(head);
    if (description != null) totalDescriptionLength += description.length;

    const finding = validateSkillFile(skillFile);
    if (finding) findings.push(finding);
  }

  if (totalDescriptionLength > MAX_TOTAL_DESCRIPTION_LENGTH) {
    findings.push({
      path: join(skillsDir, "SKILL_DESCRIPTIONS"),
      problems: [
        `frontmatter description total too long (${totalDescriptionLength} > ${MAX_TOTAL_DESCRIPTION_LENGTH} characters)`,
      ],
    });
  }

  return findings;
}

export function printSkillFindings(skillsDir, findings) {
  for (const finding of findings) {
    console.log(relative(skillsDir, finding.path));
    for (const problem of finding.problems) console.log(`  - ${problem}`);
  }

  if (findings.length > 0) {
    console.log("");
    console.log(`${findings.length} skill(s) failed anatomy validation`);
  } else {
    console.log("all skills conform to the anatomy");
  }
}

function pluginSkillsDir(skillsDir) {
  return resolve(skillsDir, "../../..", "plugin/skills");
}

function repoRootForSkillsDir(skillsDir) {
  return resolve(skillsDir, "../../..");
}

function walkFiles(root) {
  const files = [];
  if (!existsSync(root)) return files;
  for (const name of readdirSync(root).sort()) {
    const path = join(root, name);
    const stat = statSync(path);
    if (stat.isDirectory()) files.push(...walkFiles(path));
    else if (stat.isFile()) files.push(path);
  }
  return files;
}

function sameBytes(left, right) {
  return readFileSync(left).equals(readFileSync(right));
}

export function validatePluginDrift(skillsDir) {
  const pluginDir = pluginSkillsDir(skillsDir);
  if (!existsSync(pluginDir) || !statSync(pluginDir).isDirectory()) return 0;

  let drift = 0;
  const skillDirs = readdirSync(skillsDir)
    .sort()
    .map((name) => join(skillsDir, name))
    .filter((path) => statSync(path).isDirectory());

  for (const skillDir of skillDirs) {
    const mirror = join(pluginDir, skillDir.split(/[\\/]/).at(-1));
    if (!existsSync(mirror) || !statSync(mirror).isDirectory()) {
      console.log(`plugin drift: ${mirror} missing -- run scripts/generate-plugin-symlinks.mjs`);
      drift += 1;
      continue;
    }

    for (const sourceFile of walkFiles(skillDir)) {
      const rel = relative(skillDir, sourceFile);
      const mirrorFile = join(mirror, rel);
      if (!existsSync(mirrorFile) || !statSync(mirrorFile).isFile()) {
        console.log(`plugin drift: ${mirrorFile} missing`);
        drift += 1;
        continue;
      }
      if (!sameBytes(sourceFile, mirrorFile)) {
        console.log(`plugin drift: ${mirrorFile} differs from ${sourceFile}`);
        drift += 1;
      }
    }

    for (const mirrorFile of walkFiles(mirror)) {
      const rel = relative(mirror, mirrorFile);
      const sourceFile = join(skillDir, rel);
      if (!existsSync(sourceFile) || !statSync(sourceFile).isFile()) {
        console.log(`plugin drift: ${mirrorFile} has no canonical source`);
        drift += 1;
      }
    }
  }

  for (const name of readdirSync(pluginDir).sort()) {
    const mirror = join(pluginDir, name);
    if (statSync(mirror).isDirectory() && !existsSync(join(skillsDir, name))) {
      console.log(`plugin drift: ${mirror} has no canonical skill`);
      drift += 1;
    }
  }

  if (drift === 0) console.log("plugin/ skill mirror in sync with source");
  else {
    console.log("");
    console.log(`${drift} plugin mirror difference(s) found`);
  }
  return drift;
}

function readJsonObject(path) {
  if (!existsSync(path) || !statSync(path).isFile()) return [null, `missing ${path}`];
  try {
    const data = JSON.parse(readFileSync(path, "utf8"));
    if (data == null || Array.isArray(data) || typeof data !== "object") {
      return [null, `${path} must contain a JSON object`];
    }
    return [data, null];
  } catch (error) {
    return [null, `${path} is not valid JSON: ${error.message.split("\n")[0]}`];
  }
}

function firstPluginEntry(marketplace) {
  if (!Array.isArray(marketplace.plugins)) return null;
  return marketplace.plugins.find((entry) => entry && typeof entry === "object" && entry.name === "abp") ?? null;
}

export function validateCodexPluginPackage(skillsDir) {
  const root = repoRootForSkillsDir(skillsDir);
  const problems = [];
  const marketplacePath = join(root, ".agents/plugins/marketplace.json");
  const manifestPath = join(root, "plugin/.codex-plugin/plugin.json");
  const hooksPath = join(root, "plugin/.codex-plugin/hooks.json");
  const claudeMarketplacePath = join(root, ".claude-plugin/marketplace.json");
  const claudeManifestPath = join(root, "plugin/.claude-plugin/plugin.json");

  const [marketplace, marketplaceProblem] = readJsonObject(marketplacePath);
  if (marketplaceProblem) problems.push(marketplaceProblem);
  else if (marketplace) {
    if (marketplace.name !== "abp") problems.push(`${marketplacePath} name must be 'abp'`);
    if (!marketplace.interface || typeof marketplace.interface !== "object") {
      problems.push(`${marketplacePath} interface must be an object`);
    } else if (marketplace.interface.displayName !== "Agent Booster Pack") {
      problems.push(`${marketplacePath} interface.displayName must be 'Agent Booster Pack'`);
    }

    const entry = firstPluginEntry(marketplace);
    if (!entry) problems.push(`${marketplacePath} must include an 'abp' plugin entry`);
    else {
      if (!entry.source || typeof entry.source !== "object") {
        problems.push(`${marketplacePath} abp.source must be an object`);
      } else {
        if (entry.source.source !== "local") problems.push(`${marketplacePath} abp.source.source must be 'local'`);
        if (entry.source.path !== "./plugin") problems.push(`${marketplacePath} abp.source.path must be './plugin'`);
      }

      if (!entry.policy || typeof entry.policy !== "object") {
        problems.push(`${marketplacePath} abp.policy must be an object`);
      } else {
        if (entry.policy.installation !== "AVAILABLE") {
          problems.push(`${marketplacePath} abp.policy.installation must be 'AVAILABLE'`);
        }
        if (entry.policy.authentication !== "ON_INSTALL") {
          problems.push(`${marketplacePath} abp.policy.authentication must be 'ON_INSTALL'`);
        }
      }

      if (entry.category !== "Coding") problems.push(`${marketplacePath} abp.category must be 'Coding'`);
    }
  }

  const [manifest, manifestProblem] = readJsonObject(manifestPath);
  if (manifestProblem) problems.push(manifestProblem);
  else if (manifest) {
    if (manifest.name !== "abp") problems.push(`${manifestPath} name must be 'abp'`);
    if (manifest.skills !== "./skills/") problems.push(`${manifestPath} skills must be './skills/'`);
    if (manifest.hooks !== "./hooks.json") problems.push(`${manifestPath} hooks must be './hooks.json'`);

    const iface = manifest.interface;
    if (!iface || typeof iface !== "object") {
      problems.push(`${manifestPath} interface must be an object`);
    } else {
      if (iface.displayName !== "Agent Booster Pack") {
        problems.push(`${manifestPath} interface.displayName must be 'Agent Booster Pack'`);
      }
      if (iface.category !== "Coding") problems.push(`${manifestPath} interface.category must be 'Coding'`);
      if (!Array.isArray(iface.capabilities) || !["Read", "Write"].every((cap) => iface.capabilities.includes(cap))) {
        problems.push(`${manifestPath} interface.capabilities must include 'Read' and 'Write'`);
      }
      if (!Array.isArray(iface.defaultPrompt) || iface.defaultPrompt.length > 3) {
        problems.push(`${manifestPath} interface.defaultPrompt must contain at most 3 prompts`);
      }
    }

    const [hooks, hooksProblem] = readJsonObject(hooksPath);
    if (hooksProblem) {
      problems.push(hooksProblem);
    } else {
      const stopGroups = hooks.hooks?.Stop;
      const stopCommands = Array.isArray(stopGroups)
        ? stopGroups.flatMap((group) => (Array.isArray(group?.hooks) ? group.hooks : []))
        : [];
      const hasSelfReviewStopHook = stopCommands.some(
        (hook) =>
          hook?.type === "command" &&
          hook.command === "node ${PLUGIN_ROOT}/scripts/self-review.mjs" &&
          hook.timeout === 5,
      );
      if (!hasSelfReviewStopHook) {
        problems.push(`${hooksPath} must register the ABP self-review Stop hook`);
      }
    }

    const [claudeMarketplace, claudeProblem] = readJsonObject(claudeMarketplacePath);
    if (!claudeProblem && claudeMarketplace) {
      const claudeVersion = claudeMarketplace.metadata?.version;
      if (manifest.version !== claudeVersion) {
        problems.push(`${manifestPath} version must match ${claudeMarketplacePath} metadata.version`);
      }
      const claudeEntry = firstPluginEntry(claudeMarketplace);
      if (!claudeEntry) {
        problems.push(`${claudeMarketplacePath} must include an 'abp' plugin entry`);
      } else {
        if (claudeEntry.source !== "./plugin") {
          problems.push(`${claudeMarketplacePath} abp.source must be './plugin'`);
        }
        if (claudeEntry.version !== claudeVersion) {
          problems.push(`${claudeMarketplacePath} abp.version must match metadata.version`);
        }
      }
    }

    const [claudeManifest, claudeManifestProblem] = readJsonObject(claudeManifestPath);
    if (!claudeManifestProblem && claudeManifest && claudeMarketplace) {
      const claudeVersion = claudeMarketplace.metadata?.version;
      if (claudeManifest.version !== claudeVersion) {
        problems.push(`${claudeManifestPath} version must match ${claudeMarketplacePath} metadata.version`);
      }
      if (manifest.version !== claudeManifest.version) {
        problems.push(`${manifestPath} version must match ${claudeManifestPath} version`);
      }
    }
  }

  if (problems.length > 0) {
    for (const problem of problems) console.log(`codex plugin: ${problem}`);
    console.log("");
    console.log(`${problems.length} codex plugin package problem(s)`);
  } else {
    console.log("codex plugin package valid");
  }
  return problems.length;
}

function writeFixture(path, text) {
  mkdirSync(dirname(path), { recursive: true });
  writeFileSync(path, text, { encoding: "utf8", flag: "w" });
}

export function runSelfTest() {
  const tmp = mkdtempSync(join(tmpdir(), "abp-skill-validator-"));
  try {
    cpSync(resolve(scriptDir(), "../agents/.agents/skills/workflow"), join(tmp, "good"), { recursive: true });
    writeFixture(
      join(tmp, "bad/SKILL.md"),
      `---
name: bad
description: >-
  This description is intentionally much too long for the skill routing surface
  because Codex and other agents load all skill descriptions before deciding
  which full skill body to request, so verbose trigger prose creates avoidable
  context pressure.
---

# Bad

Per Rich Hickey, prefer values over places.

## Overview

Stuff.

## Common Rationalizations
| Excuse | Reality |
|---|---|
| "Should work" | It might not. |

## Red Flags
- "Probably fine"
`,
    );

    const longDescription = "x".repeat(100);
    for (let index = 0; index < 21; index += 1) {
      writeFixture(
        join(tmp, `budget-${index}/SKILL.md`),
        `---
name: budget-${index}
description: ${longDescription}
---

# Budget ${index}

## When to Use
- trigger

## When NOT to Use
- other

## Verification
- [ ] check
`,
      );
    }

    const findings = validateSkills(tmp);
    const rendered = [
      ...findings.map((finding) => relative(tmp, finding.path)),
      ...findings.flatMap((finding) => finding.problems),
    ].join("\n");
    for (const expected of [
      "bad/SKILL.md",
      "When to Use",
      "Verification",
      "description too long",
      "per <expert>",
      "Common Rationalizations",
      "Red Flags",
      "positive Tripwires bullets",
      "description total too long",
    ]) {
      if (!rendered.includes(expected)) {
        console.error(`self-test failed: missing ${JSON.stringify(expected)} in output`);
        console.error(rendered);
        return 1;
      }
    }

    if (findings.some((finding) => relative(tmp, finding.path) === "good/SKILL.md")) {
      console.error("self-test failed: good skill flagged");
      console.error(rendered);
      return 1;
    }

    console.log("self-test ok");
    return 0;
  } finally {
    rmSync(tmp, { recursive: true, force: true });
  }
}

export function main(argv = process.argv.slice(2)) {
  if (argv.includes("-h") || argv.includes("--help")) {
    console.log(`Usage: node scripts/validate-skill-anatomy.mjs [skills_dir] [--self-test]

Validate SKILL.md frontmatter, required sections, and plugin drift.`);
    return 0;
  }
  if (argv.includes("--self-test")) return runSelfTest();

  const skillsDir = resolve(argv.find((arg) => !arg.startsWith("-")) ?? defaultSkillsDir());
  if (!existsSync(skillsDir) || !statSync(skillsDir).isDirectory()) {
    console.error(`not a directory: ${skillsDir}`);
    return 2;
  }

  const findings = validateSkills(skillsDir);
  printSkillFindings(skillsDir, findings);
  const drift = validatePluginDrift(skillsDir);
  const codexPluginProblems = validateCodexPluginPackage(skillsDir);
  return findings.length || drift || codexPluginProblems ? 1 : 0;
}

if (import.meta.url === `file://${process.argv[1]}`) {
  process.exitCode = main();
}
